app/models/concerns/raceable.rb

```diff
@@ -0,0 +1,71 @@
+module Raceable
+  extend ActiveSupport::Concern
+
+  included do
+    OPEN = 'Open Entry'.freeze
+    IN_PROGRESS = 'In Progress'.freeze
+    ENDED = 'Ended'.freeze
+
+    belongs_to :owner, foreign_key: :user_id, class_name: 'User'
+    has_many :entrants, as: :raceable, dependent: :destroy
+    has_many :users, through: :entrants
+    has_one :chat_room, dependent: :destroy
+    has_many :chat_messages, through: :chat_room
+
+    has_secure_token :auth_token
+
+    def public?
+      listed? && !invite?
+    end
+
+    def invite_only?
+      listed? && invite?
+    end
+
+    def secret?
+      !listed?
+    end
+
+    def started?
+      started_at.present?
+    end
+
+    def in_progress?
+      started? && !finished?
+    end
+
+    def finished?
+      entrants.all? { |entrant| entrant.finished? || entrant.forfeited? }
+    end
+
+    def entrant_for_user(user)
+      entrants.find_by(user_id: user.id)
+    end
+
+    def visibility!(type)
+      # When given an invalid type, the column defaults will make the race public
```

Comment:
imo invalid types should error, otherwise we might write some code that slips through the cracks into production because it mostly looks like it works. Even if it's just an uncaught exception that bubbles up to cause a 500, that's better than a 200 that does the wrong thing.

Reply:
For stuff that can accept user input, is returning `nil` acceptable if it doesn't find anything that matches?  For internal only things I agree an error is the way to go.